diff -Naur ./CAmkES/CMakeLists.txt ../HAMR_Simple_V4-patched/CAmkES/CMakeLists.txt
--- ./CAmkES/CMakeLists.txt	2021-06-21 14:43:37.032029340 -0500
+++ ../HAMR_Simple_V4-patched/CAmkES/CMakeLists.txt	2021-06-21 14:28:20.365069008 -0500
@@ -58,23 +58,24 @@
 DeclareCAmkESComponent(AttestationManager_Impl_SW_AttestationManager_AttestationManager
   SOURCES components/AttestationManager_Impl_SW_AttestationManager_AttestationManager/src/sb_AttestationManager_Impl_ffi.c components/AttestationManager_Impl_SW_AttestationManager_AttestationManager/src/heli_am.S components/AttestationManager_Impl_SW_AttestationManager_AttestationManager/src/sb_AttestationManager_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
   INCLUDES components/AttestationManager_Impl_SW_AttestationManager_AttestationManager/includes/ types/includes
-  LIBS AttestationManager_Impl_SW_AttestationManager_AttestationManager
+  LIBS heli_am_c AttestationManager_Impl_SW_AttestationManager_AttestationManager
+  LD_FLAGS -L ${CMAKE_CURRENT_LIST_DIR}/components/AttestationManager_Impl_SW_AttestationManager_AttestationManager/src
 )
 
 DeclareCAmkESComponent(AttestationGate_Impl_SW_AttestationGate_AttestationGate
-  SOURCES components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl_ffi.c components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/AttestationGate.S components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
+  SOURCES components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl_ffi.c components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/basis_ffi.c components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/AttestationGate.S components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
   INCLUDES components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/includes/ types/includes
   LIBS AttestationGate_Impl_SW_AttestationGate_AttestationGate
 )
 
 DeclareCAmkESComponent(Filter_Impl_SW_Filter_Filter
-  SOURCES components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl_ffi.c components/Filter_Impl_SW_Filter_Filter/src/Filter.S components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
+  SOURCES components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl_ffi.c components/Filter_Impl_SW_Filter_Filter/src/basis_ffi.c components/Filter_Impl_SW_Filter_Filter/src/Filter.S components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
   INCLUDES components/Filter_Impl_SW_Filter_Filter/includes/ types/includes
   LIBS Filter_Impl_SW_Filter_Filter
 )
 
 DeclareCAmkESComponent(Monitor_Impl_SW_Monitor_Monitor
-  SOURCES components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl_ffi.c components/Monitor_Impl_SW_Monitor_Monitor/src/Monitor.S components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
+  SOURCES components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl_ffi.c components/Monitor_Impl_SW_Monitor_Monitor/src/basis_ffi.c components/Monitor_Impl_SW_Monitor_Monitor/src/Monitor.S components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
   INCLUDES components/Monitor_Impl_SW_Monitor_Monitor/includes/ types/includes
   LIBS Monitor_Impl_SW_Monitor_Monitor
 )
@@ -82,7 +83,8 @@
 DeclareCAmkESComponent(AttestationTester_Impl_SW_AttestationTester_AttestationTester
   SOURCES components/AttestationTester_Impl_SW_AttestationTester_AttestationTester/src/sb_AttestationTester_Impl_ffi.c components/AttestationTester_Impl_SW_AttestationTester_AttestationTester/src/user_am.S components/AttestationTester_Impl_SW_AttestationTester_AttestationTester/src/sb_AttestationTester_Impl.c types/src/sb_queue_union_art_DataContent_1.c types/src/sb_queue_union_art_DataContent_1.c
   INCLUDES components/AttestationTester_Impl_SW_AttestationTester_AttestationTester/includes/ types/includes
-  LIBS AttestationTester_Impl_SW_AttestationTester_AttestationTester
+  LIBS heli_am_c AttestationTester_Impl_SW_AttestationTester_AttestationTester
+  LD_FLAGS -L ${CMAKE_CURRENT_LIST_DIR}/components/AttestationTester_Impl_SW_AttestationTester_AttestationTester/src
 )
 
 DeclareCAmkESRootserver(MissionComputer_Impl_Instance.camkes)
diff -Naur ./CAmkES/components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl_ffi.c ../HAMR_Simple_V4-patched/CAmkES/components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl_ffi.c
--- ./CAmkES/components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl_ffi.c	2021-06-21 14:43:37.032029340 -0500
+++ ../HAMR_Simple_V4-patched/CAmkES/components/AttestationGate_Impl_SW_AttestationGate_AttestationGate/src/sb_AttestationGate_Impl_ffi.c	2021-06-21 14:09:50.591252567 -0500
@@ -158,32 +158,6 @@
  * Required by the FFI framework
  */
 
-void ffiwrite (unsigned char *parameter, long parameterSizeBytes, unsigned char *output, long outputSizeBytes){
-}
-
-void cml_exit(int arg) {
-  #ifdef DEBUG_FFI
-  {
-    fprintf(stderr,"GCNum: %d, GCTime(us): %ld\n",numGC,microsecs);
-  }
-  #endif
-  exit(arg);
-}
-
-// convert big-endian 32-bit float to little-endian 64 bit double
-void ffifloat2double(unsigned char *parameter, long parameterSizeBytes,
-                     unsigned char *output,    long outputSizeBytes) {
-  char bytes [4];
-  assert (4 == parameterSizeBytes);
-  bytes[3] = parameter[0];
-  bytes[2] = parameter[1];
-  bytes[1] = parameter[2];
-  bytes[0] = parameter[3];
-
-  double result = *((float*)bytes);
-  memcpy(output, (unsigned char*) &result, sizeof(double));
-}
-
 void ffiraw_print_buffer(unsigned char *parameter, long parameterSizeBytes, unsigned char *output, long outputSizeBytes) {
 
   int elideAfter = 20; // stop printing after this many consecutive 0's seen
diff -Naur ./CAmkES/components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl_ffi.c ../HAMR_Simple_V4-patched/CAmkES/components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl_ffi.c
--- ./CAmkES/components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl_ffi.c	2021-06-21 14:43:37.064029841 -0500
+++ ../HAMR_Simple_V4-patched/CAmkES/components/Filter_Impl_SW_Filter_Filter/src/sb_Filter_Impl_ffi.c	2021-06-21 14:09:50.591252567 -0500
@@ -148,32 +148,6 @@
  * Required by the FFI framework
  */
 
-void ffiwrite (unsigned char *parameter, long parameterSizeBytes, unsigned char *output, long outputSizeBytes){
-}
-
-void cml_exit(int arg) {
-  #ifdef DEBUG_FFI
-  {
-    fprintf(stderr,"GCNum: %d, GCTime(us): %ld\n",numGC,microsecs);
-  }
-  #endif
-  exit(arg);
-}
-
-// convert big-endian 32-bit float to little-endian 64 bit double
-void ffifloat2double(unsigned char *parameter, long parameterSizeBytes,
-                     unsigned char *output,    long outputSizeBytes) {
-  char bytes [4];
-  assert (4 == parameterSizeBytes);
-  bytes[3] = parameter[0];
-  bytes[2] = parameter[1];
-  bytes[1] = parameter[2];
-  bytes[0] = parameter[3];
-
-  double result = *((float*)bytes);
-  memcpy(output, (unsigned char*) &result, sizeof(double));
-}
-
 void ffiraw_print_buffer(unsigned char *parameter, long parameterSizeBytes, unsigned char *output, long outputSizeBytes) {
 
   int elideAfter = 20; // stop printing after this many consecutive 0's seen
diff -Naur ./CAmkES/components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl_ffi.c ../HAMR_Simple_V4-patched/CAmkES/components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl_ffi.c
--- ./CAmkES/components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl_ffi.c	2021-06-21 14:43:37.064029841 -0500
+++ ../HAMR_Simple_V4-patched/CAmkES/components/Monitor_Impl_SW_Monitor_Monitor/src/sb_Monitor_Impl_ffi.c	2021-06-21 14:09:50.591252567 -0500
@@ -155,32 +155,6 @@ 
  * Required by the FFI framework
  */
 
-void ffiwrite (unsigned char *parameter, long parameterSizeBytes, unsigned char *output, long outputSizeBytes){
-}
-
-void cml_exit(int arg) {
-  #ifdef DEBUG_FFI
-  {
-    fprintf(stderr,"GCNum: %d, GCTime(us): %ld\n",numGC,microsecs);
-  }
-  #endif
-  exit(arg);
-}
-
-// convert big-endian 32-bit float to little-endian 64 bit double
-void ffifloat2double(unsigned char *parameter, long parameterSizeBytes,
-                     unsigned char *output,    long outputSizeBytes) {
-  char bytes [4];
-  assert (4 == parameterSizeBytes);
-  bytes[3] = parameter[0];
-  bytes[2] = parameter[1];
-  bytes[1] = parameter[2];
-  bytes[0] = parameter[3];
-
-  double result = *((float*)bytes);
-  memcpy(output, (unsigned char*) &result, sizeof(double));
-}
-
 void ffiraw_print_buffer(unsigned char *parameter, long parameterSizeBytes, unsigned char *output, long outputSizeBytes) {
 
   int elideAfter = 20; // stop printing after this many consecutive 0's seen
